File: testing-unit-fs-mocks/WebsiteStackTests.fs
// Copyright 2016-2020, Pulumi Corporation

namespace UnitTesting

open System
open System.Collections.Immutable
open System.Threading.Tasks
open NUnit.Framework
open Pulumi
open Pulumi.Azure.Core
open Pulumi.Azure.Storage
open Pulumi.Testing
open FluentAssertions

[<TestFixture>]
type WebserverStackTests() =
    let runTest(): ImmutableArray<Resource> =
        let options = new TestOptions(IsPreview = Nullable<bool> false)
        Deployment.TestAsync<WebsiteStack>(new Mocks(), options)
        |> Async.AwaitTask
        |> Async.RunSynchronously
        
    let getValue(output: Output<'a>): 'a =
        let tcs = new TaskCompletionSource<'a>()
        output.Apply(fun v -> tcs.SetResult(v); v) |> ignore
        tcs.Task.Result

    [<Test>]
    member this.SingleResourceGroupExists() =
        let resources = runTest()

        let resourceGroupCount = resources.OfType<ResourceGroup>() |> Seq.length
        resourceGroupCount.Should().Be(1, "a single resource group is expected") |> ignore    

    [<Test>]
    member this.ResourceGroupHasEnvironmentTag() =
        let resources = runTest()
        let resourceGroup = resources.OfType<ResourceGroup>() |> Seq.head

        let tags = getValue resourceGroup.Tags
        tags.Should().NotBeNull("Tags must be defined") |> ignore
        tags.Should().ContainKey("Environment", null) |> ignore

    [<Test>]
    member this.StorageAccountBelongsToResourceGroup() =
        let resources = runTest()
        let storageAccount = resources.OfType<Account>() |> Seq.tryHead |> Option.toObj
        storageAccount.Should().NotBeNull("Storage account not found") |> ignore
        
        let resourceGroupName = getValue storageAccount.ResourceGroupName
        resourceGroupName.Should().Be("www-prod-rg", null) |> ignore

    [<Test>]
    member this.UploadsTwoFiles() =
        let resources = runTest()
        let filesCount = resources.OfType<Blob>() |> Seq.length
        filesCount.Should().Be(2, "Should have uploaded files from `wwwroot`") |> ignore
        
    [<Test>]
    member this.StackExportsWebsiteUrl() =
        let resources = runTest()
        let stack = resources.OfType<WebsiteStack>() |> Seq.head
    
        let endpoint = getValue stack.Endpoint
        endpoint.Should().Be("https://wwwprodsa.web.core.windows.net", null) |> ignore

File: testing-unit-cs-mocks/WebsiteStackTests.cs
// Copyright 2016-2020, Pulumi Corporation

using System.Collections.Immutable;
using System.Linq;
using System.Threading.Tasks;
using FluentAssertions;
using NUnit.Framework;
using Pulumi;
using Pulumi.Testing;
using Pulumi.AzureNative.Resources;
using Pulumi.AzureNative.Storage;

namespace UnitTesting
{
	/// <summary>
	/// Unit testing examples.
	/// </summary>
	[TestFixture]
	public class WebsiteStackTests
	{
		private static Task<ImmutableArray<Pulumi.Resource>> TestAsync()
		{
			return Pulumi.Deployment.TestAsync<WebsiteStack>(new Mocks(), new TestOptions {IsPreview = false});
		}
		
		[Test]
		public async Task SingleResourceGroupExists()
		{
			var resources = await TestAsync();

			var resourceGroups = resources.OfType<ResourceGroup>().ToList();
			resourceGroups.Count.Should().Be(1, "a single resource group is expected");
		}
		
		[Test]
		public async Task ResourceGroupHasEnvironmentTag()
		{
			var resources = await TestAsync();
			var resourceGroup = resources.OfType<ResourceGroup>().First();
		
			var tags = await resourceGroup.Tags.GetValueAsync();
			tags.Should().NotBeNull("Tags must be defined");
			tags.Should().ContainKey("Environment");
		}
		
		[Test]
		public async Task StorageAccountExists()
		{
			var resources = await TestAsync();
			var storageAccounts = resources.OfType<StorageAccount>();
			var storageAccount = storageAccounts.SingleOrDefault();
			storageAccount.Should().NotBeNull("Storage account not found");
		}
		
		[Test]
		public async Task UploadsTwoFiles()
		{
			var resources = await TestAsync();
			var files = resources.OfType<Blob>().ToList();
			files.Count.Should().Be(2, "Should have uploaded files from `wwwroot`");
		}
		
		[Test]
		public async Task StackExportsWebsiteUrl()
		{
			var resources = await TestAsync();
			var stack = resources.OfType<WebsiteStack>().First();
			
			var endpoint = await stack.Endpoint.GetValueAsync();
			endpoint.Should().Be("https://wwwprodsa.web.core.windows.net");
		}
	}
}
